src/firefox.py

```diff
@@ -0,0 +1,135 @@
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date_gt_3(soup: BeautifulSoup) -> Tuple[str, str]:
+    """ Version matching for firefox versions >= 3.0 """
+    release_info = soup.select('div#main-feature p em')[0].get_text()
+
+    # get version
+    version_match = re.search(VERSION_REGEX, release_info)
+    if version_match is None:
+        raise UnsupportedReleasePageError("Unable to find version")
+    version = version_match.group()
+
+    # get date
+    unformatted_date_match = re.search(DATE_REGEX, release_info)
+    if unformatted_date_match is None:
+        raise UnsupportedReleasePageError("Unable to find date")
+    unformatted_date = unformatted_date_match.group()
+    date = format_date(unformatted_date)
+
+    return (version, date)
+
+def get_version_and_date(soup: BeautifulSoup, release_page: str) -> Tuple[str, str]:
+    """ Get version and date from the given release page """
+    functions = [get_version_and_date_gt_28, get_version_and_date_gt_10, get_version_and_date_gt_3]
+
+    # Note: firefox release pages for versions <3.0 don't include release dates so we
+    # can't match these versions for now.
+    # example: https://www.mozilla.org/en-US/firefox/2.0/releasenotes/
+    for function in functions:
+        try:
+            return function(soup)
+        except (UnsupportedReleasePageError, AttributeError, IndexError):
+            pass
+
+    raise UnsupportedReleasePageError("Unable to find version and date for %s" % release_page)
+
+def make_bs_request(url: str) -> BeautifulSoup:
+    """ Make a request to the given url and return a BeautifulSoup object """
+    headers = {"user-agent": "mozilla"}
+    req = urllib.request.Request(url, headers=headers)
+    res = urllib.request.urlopen(req, timeout=5)
+    return BeautifulSoup(res.read(), features="html5lib")
+
+def fetch_releases():
+    releases = {}
+    soup = make_bs_request(URL)
+
+    ff_releases = soup.find_all("ol", class_="c-release-list")
+    for p in ff_releases[0].find_all("a")[::-1]:
+        release_page = requests.compat.urljoin(URL, p.get("href"))
+        rp_soup = make_bs_request(release_page)
+
+        try:
+            (version, date) = get_version_and_date(rp_soup, release_page)
+            print("%s: %s" % (version, date))
+            releases[version] = date
+        except UnsupportedReleasePageError:
+            print("Unsupported release page: %s" % release_page)
+
+    return releases
+
+def main():
+    print(f"::group::{PRODUCT}")
+
+    releases = fetch_releases()
+    with open(f"releases/{PRODUCT}.json", "w") as f:
+        f.write(json.dumps(
+            # sort by date desc
+            dict(sorted(releases.items(), key=lambda e: e[1], reverse=True)),
```

Comment:
The sort algorithm must consider both dates and versions, or the sort will not be stable if two versions are released the same day (i.e. two execution with no updated versions could generate different files). Had this issue with java.py : #64.

Reply:
Fixed.